File: Windows/src/IwashiScope.Core/Calculations/Cie2006Chromaticity.cs
using IwashiScope.Core.Models;

namespace IwashiScope.Core.Calculations;

public readonly record struct PhysiologicalPoint(double X, double Y)
{
    public bool IsFinite => double.IsFinite(X) && double.IsFinite(Y);
}

public enum PhysiologicalCone { Long, Medium, Short }
public enum ConfusionColorMode { C, P, D, T }
public enum ChromaticityError { InsufficientSpectrum, InvalidSpectrum, ZeroSignal }
public sealed record ChromaticitySegment(PhysiologicalPoint Start, PhysiologicalPoint End)
{
    public PhysiologicalPoint At(double t) => new(Start.X + t * (End.X - Start.X), Start.Y + t * (End.Y - Start.Y));
}
public sealed record PhysiologicalMeasurement(Vector3 Lms, Vector3 WhiteLms, Vector3 XyzF, PhysiologicalPoint Point, WavelengthRange Range);

/// <summary>Separate D50 reference display path. Existing measurement XYZ/Lab and appearance calculations are unchanged.</summary>
public static class Cie2006Chromaticity
{
    public static IReadOnlyList<double> Sensitivities(PhysiologicalCone cone) => cone switch
    {
        PhysiologicalCone.Long => Cie2006LmsData.Long,
        PhysiologicalCone.Medium => Cie2006LmsData.Medium,
        _ => Cie2006LmsData.Short,
    };

    // CIE 170-2:2015 / Stockman (2019), Eq. 4. Raw channel gains are retained.
    public static Vector3 XyzF(Vector3 lms) => new(
        1.94735469 * lms.First - 1.41445123 * lms.Second + 0.36476327 * lms.Third,
        0.68990272 * lms.First + 0.34832189 * lms.Second, 1.93485343 * lms.Third);

    public static PhysiologicalPoint? Point(Vector3 xyz)
    {
        var sum = xyz.First + xyz.Second + xyz.Third;
        if (!xyz.IsFinite || !double.IsFinite(sum) || sum == 0) return null;
        var result = new PhysiologicalPoint(xyz.First / sum, xyz.Second / sum);
        return result.IsFinite ? result : null;
    }

    public static double Sensitivity(IReadOnlyList<double> values, double wavelength)
    {
        var position = (wavelength - Cie2006LmsData.FirstWavelength) / Cie2006LmsData.Interval;
        if (!double.IsFinite(position) || position < 0 || position > values.Count - 1) return 0;
        var i = (int)Math.Floor(position);
        return values[i] + (values[Math.Min(i + 1, values.Count - 1)] - values[i]) * (position - i);
    }

    public static Vector3? SpectralLms(double wavelength) => double.IsFinite(wavelength) && wavelength >= 390 && wavelength <= 830
        ? new(Sensitivity(Cie2006LmsData.Long, wavelength), Sensitivity(Cie2006LmsData.Medium, wavelength), Sensitivity(Cie2006LmsData.Short, wavelength)) : null;

    public static PhysiologicalPoint CopunctalPoint(PhysiologicalCone cone) => Point(XyzF(cone switch
    {
        PhysiologicalCone.Long => new(1, 0, 0), PhysiologicalCone.Medium => new(0, 1, 0), _ => new(0, 0, 1),
    }))!.Value;

    public static ChromaticitySegment? ConfusionLine(PhysiologicalPoint point, PhysiologicalCone cone)
    {
        if (!point.IsFinite) return null;
        var pole = CopunctalPoint(cone);
        var dx = pole.X - point.X; var dy = pole.Y - point.Y;
        if (Math.Sqrt(dx * dx + dy * dy) <= 1e-12) return null;
        var lower = double.NegativeInfinity; var upper = double.PositiveInfinity;
        foreach (var (origin, direction, maximum) in new[] { (point.X, dx, 0.8), (point.Y, dy, 0.9) })
        {
            if (Math.Abs(direction) < 1e-12) { if (origin < 0 || origin > maximum) return null; }
            else
            {
                var t1 = -origin / direction; var t2 = (maximum - origin) / direction;
                lower = Math.Max(lower, Math.Min(t1, t2)); upper = Math.Min(upper, Math.Max(t1, t2));
            }
        }
        return double.IsFinite(lower) && double.IsFinite(upper) && lower < upper
            ? new(new(point.X + lower * dx, point.Y + lower * dy), new(point.X + upper * dx, point.Y + upper * dy)) : null;
    }

    public static PhysiologicalMeasurement? Measure(SpotMeasurement? measurement, out ChromaticityError? error)
    {
        error = ChromaticityError.InsufficientSpectrum;
        if (measurement?.Mode != MeasurementMode.Reflectance || measurement.Spectrum.Count < 2) return null;
        error = ChromaticityError.InvalidSpectrum;
        if (measurement.Spectrum.Any(s => !double.IsFinite(s.Wavelength) || !double.IsFinite(s.Value))) return null;
        var reflectance = measurement.Spectrum.OrderBy(s => s.Wavelength).Select(s => s with { Value = Math.Max(0, s.Value) / 100 }).ToArray();
        if (reflectance.Zip(reflectance.Skip(1)).Any(p => p.First.Wavelength >= p.Second.Wavelength)) return null;
        var illuminant = CieReferenceIlluminants.Samples(CieReferenceIlluminant.D50);
        var lower = Math.Max(390, Math.Max(reflectance[0].Wavelength, illuminant[0].Wavelength));
        var upper = Math.Min(830, Math.Min(reflectance[^1].Wavelength, illuminant[^1].Wavelength));
        if (lower >= upper) { error = ChromaticityError.InsufficientSpectrum; return null; }
        var integrals = Enum.GetValues<PhysiologicalCone>().Select(c => Integrate(c, reflectance, illuminant, lower, upper)).ToArray();
        var whiteY = XyzF(new(integrals[0].White, integrals[1].White, integrals[2].White)).Second;
        if (!double.IsFinite(whiteY) || whiteY <= 0) return null;
        var normalization = 100 / whiteY;
        var lms = new Vector3(integrals[0].Object * normalization, integrals[1].Object * normalization, integrals[2].Object * normalization);
        var xyz = XyzF(lms);
        if (!xyz.IsFinite) return null;
        if (xyz.First == 0 && xyz.Second == 0 && xyz.Third == 0) { error = ChromaticityError.ZeroSignal; return null; }
        if (Point(xyz) is not { } point) return null;
        error = null;
        var whiteLms = new Vector3(integrals[0].White * normalization, integrals[1].White * normalization, integrals[2].White * normalization);
        return new(lms, whiteLms, xyz, point, new(lower, upper));
    }

    private static (double Object, double White) Integrate(PhysiologicalCone cone, IReadOnlyList<SpectralSample> reflectance,
        IReadOnlyList<SpectralSample> illuminant, double lower, double upper)
    {
        var values = Sensitivities(cone);
        upper = Math.Min(upper, 390 + (values.Count - 1) * 5);
        if (lower >= upper) return (0, 0);
        var nodes = new[] { lower, upper }.Concat(reflectance.Select(s => s.Wavelength)).Concat(illuminant.Select(s => s.Wavelength))
            .Concat(Enumerable.Range(0, values.Count).Select(i => 390d + i * 5)).Where(w => w >= lower && w <= upper).Distinct().Order().ToArray();
        double obj = 0, white = 0;
        for (var i = 0; i < nodes.Length - 1; i++)
        {
            var a = nodes[i]; var b = nodes[i + 1];
            // The product of three linear factors is cubic: Simpson is exact between merged knots.
            foreach (var (w, weight) in new[] { (a, 1d), ((a + b) / 2, 4d), (b, 1d) })
            {
                var weighted = Interpolate(illuminant, w) * Sensitivity(values, w) * (b - a) * weight / 6;
                white += weighted; obj += weighted * Interpolate(reflectance, w);
            }
        }
        return (obj, white);
    }

    internal static double Interpolate(IReadOnlyList<SpectralSample> values, double w)
    {
        var lower = 0; var upper = values.Count - 1;
        while (lower + 1 < upper) { var mid = (lower + upper) / 2; if (values[mid].Wavelength <= w) lower = mid; else upper = mid; }
        var fraction = (w - values[lower].Wavelength) / (values[upper].Wavelength - values[lower].Wavelength);
        return values[lower].Value + (values[upper].Value - values[lower].Value) * fraction;
    }
}

/// <summary>Peak-normalized display curves. Never used for colorimetry or y-axis scale selection.</summary>
public static class Cie2006LmsReference
{
    public const double StrokeOpacity = 0.4;
    public sealed record Curve(PhysiologicalCone Cone, IReadOnlyList<SpectralSample> Samples);
    public static readonly IReadOnlyList<Curve> FullCurves = Enum.GetValues<PhysiologicalCone>().Select(c =>
    {
        var values = Cie2006Chromaticity.Sensitivities(c); var peak = values.Max();
        return new Curve(c, values.Select((v, i) => new SpectralSample(i, 390 + i * 5, v / peak)).ToArray());
    }).ToArray();

    public static IReadOnlyList<Curve> Curves(double start, double end)
    {
        if (!double.IsFinite(start) || !double.IsFinite(end) || start >= end) return [];
        return FullCurves.Select(c =>
        {
            var lower = Math.Max(start, c.Samples[0].Wavelength); var upper = Math.Min(end, c.Samples[^1].Wavelength);
            return lower >= upper ? null : new Curve(c.Cone, new[] { lower }.Concat(c.Samples.Where(s => s.Wavelength > lower && s.Wavelength < upper).Select(s => s.Wavelength)).Append(upper)
                .Select((w, i) => new SpectralSample(i, w, Cie2006Chromaticity.Interpolate(c.Samples, w))).ToArray());
        }).OfType<Curve>().ToArray();
    }
}

File: IwashiScopePackage/Sources/IwashiScopeFeature/Models/CIE2006Chromaticity.swift
import Foundation

struct PhysiologicalChromaticityPoint: Equatable, Sendable {
    let x: Double
    let y: Double
    var isFinite: Bool { x.isFinite && y.isFinite }
}

enum PhysiologicalCone: String, CaseIterable, Identifiable, Sendable {
    case long = "L"
    case medium = "M"
    case short = "S"
    var id: Self { self }

    var label: String {
        switch self {
        case .long: "L (P)"
        case .medium: "M (D)"
        case .short: "S (T)"
        }
    }

    var sensitivities: [Double] {
        switch self {
        case .long: CIE2006LMSData.long
        case .medium: CIE2006LMSData.medium
        case .short: CIE2006LMSData.short
        }
    }

    var unitResponse: Vector3 {
        switch self {
        case .long: Vector3(first: 1, second: 0, third: 0)
        case .medium: Vector3(first: 0, second: 1, third: 0)
        case .short: Vector3(first: 0, second: 0, third: 1)
        }
    }
}

enum CIE2006ChromaticityError: Error, Equatable, LocalizedError {
    case invalidSpectrum
    case insufficientSpectrum
    case zeroSignal

    var errorDescription: String? {
        switch self {
        case .invalidSpectrum:
            String(localized: "色度を計算できないスペクトルです。")
        case .insufficientSpectrum:
            String(localized: "色度の計算に必要な反射スペクトルがありません。")
        case .zeroSignal:
            String(localized: "反射光がゼロのため、色度を定義できません。")
        }
    }
}

/// CIE 2006 2° energy fundamentals → CIE 2015 XYZ_F, not CIE 1931 XYZ.
/// Matrix: CIE 170-2:2015; Stockman (2019), "Cone fundamentals and CIE standards", Eq. 4.
/// Uses the original tabulated channel scales, never CIE2006LMSReference's plot normalization.
enum CIE2006Chromaticity {
    struct LocusSample: Equatable, Sendable {
        let wavelength: Double
        let point: PhysiologicalChromaticityPoint
    }

    struct LineSegment: Equatable, Sendable {
        let start: PhysiologicalChromaticityPoint
        let end: PhysiologicalChromaticityPoint
    }

    struct MeasurementResult: Equatable, Sendable {
        let lms: Vector3
        /// Perfect D50 reflector over the SAME measured wavelength range and Y_F=100 scale.
        let whiteLMS: Vector3
        let xyzF: Vector3
        let point: PhysiologicalChromaticityPoint
        let wavelengthRange: WavelengthRange
    }

    static let xRange = 0.0...0.8
    static let yRange = 0.0...0.9

    static func xyzF(from lms: Vector3) -> Vector3 {
        Vector3(
            first: 1.94735469 * lms.first - 1.41445123 * lms.second + 0.36476327 * lms.third,
            second: 0.68990272 * lms.first + 0.34832189 * lms.second,
            third: 1.93485343 * lms.third
        )
    }

    static func point(from xyzF: Vector3) -> PhysiologicalChromaticityPoint? {
        let sum = xyzF.first + xyzF.second + xyzF.third
        // A negative sum is valid for a virtual copunctal point (not a physical stimulus).
        guard [xyzF.first, xyzF.second, xyzF.third, sum].allSatisfy(\.isFinite), sum != 0 else {
            return nil
        }
        let point = PhysiologicalChromaticityPoint(x: xyzF.first / sum, y: xyzF.second / sum)
        return point.isFinite ? point : nil
    }

    static func spectralLMS(at wavelength: Double) -> Vector3? {
        guard wavelength.isFinite, (390.0...830.0).contains(wavelength) else { return nil }
        return Vector3(
            first: sensitivity(CIE2006LMSData.long, at: wavelength),
            second: sensitivity(CIE2006LMSData.medium, at: wavelength),
            third: sensitivity(CIE2006LMSData.short, at: wavelength)
        )
    }

    static let spectralLocus: [LocusSample] = stride(from: 390.0, through: 830.0, by: 5.0).compactMap {
        wavelength in
        guard let lms = spectralLMS(at: wavelength), let point = point(from: xyzF(from: lms)) else {
            return nil
        }
        return LocusSample(wavelength: wavelength, point: point)
    }

    static func copunctalPoint(for cone: PhysiologicalCone) -> PhysiologicalChromaticityPoint {
        let xyz = xyzF(from: cone.unitResponse)
        let sum = xyz.first + xyz.second + xyz.third
        return PhysiologicalChromaticityPoint(x: xyz.first / sum, y: xyz.second / sum)
    }

    /// Infinite confusion line clipped to the chart's rectangle. The view additionally
    /// clips it to the spectral locus, so virtual copunctal points need not be on screen.
    static func confusionLine(
        through point: PhysiologicalChromaticityPoint, cone: PhysiologicalCone
    ) -> LineSegment? {
        guard point.isFinite else { return nil }
        let pole = copunctalPoint(for: cone)
        let dx = pole.x - point.x
        let dy = pole.y - point.y
        guard hypot(dx, dy) > 1e-12 else { return nil }
        var lower = -Double.infinity
        var upper = Double.infinity
        for (origin, direction, range) in [(point.x, dx, xRange), (point.y, dy, yRange)] {
            if abs(direction) < 1e-12 {
                guard range.contains(origin) else { return nil }
            } else {
                let t1 = (range.lowerBound - origin) / direction
                let t2 = (range.upperBound - origin) / direction
                lower = max(lower, min(t1, t2))
                upper = min(upper, max(t1, t2))
            }
        }
        guard lower.isFinite, upper.isFinite, lower < upper else { return nil }
        return LineSegment(
            start: .init(x: point.x + lower * dx, y: point.y + lower * dy),
            end: .init(x: point.x + upper * dx, y: point.y + upper * dy)
        )
    }

    /// D50 baseline only. No selected illuminant, adaptation, RGB or instrument XYZ enters this path.
    static func measuredResult(for measurement: SpotMeasurement) throws -> MeasurementResult {
        guard measurement.mode == .reflectance, measurement.spectrum.count >= 2 else {
            throw CIE2006ChromaticityError.insufficientSpectrum
        }
        guard measurement.spectrum.allSatisfy({ $0.wavelength.isFinite && $0.value.isFinite }) else {
            throw CIE2006ChromaticityError.invalidSpectrum
        }
        // Preserve the measurement; floor negative noise only in this integration's working copy.
        let reflectance = measurement.spectrum.sorted { $0.wavelength < $1.wavelength }.map {
            SpectralSample(id: $0.id, wavelength: $0.wavelength, value: max(0, $0.value) / 100)
        }
        guard zip(reflectance, reflectance.dropFirst()).allSatisfy({ $0.wavelength < $1.wavelength }) else {
            throw CIE2006ChromaticityError.invalidSpectrum
        }
        let illuminant = CIEReferenceIlluminant.d50.samples
        let lower = max(390, reflectance[0].wavelength, illuminant[0].wavelength)
        let upper = min(830, reflectance[reflectance.count - 1].wavelength, illuminant[illuminant.count - 1].wavelength)
        guard lower < upper else { throw CIE2006ChromaticityError.insufficientSpectrum }

        let integrals = PhysiologicalCone.allCases.map {
            integrate(cone: $0, reflectance: reflectance, illuminant: illuminant, range: lower...upper)
        }
        let white = Vector3(first: integrals[0].white, second: integrals[1].white, third: integrals[2].white)
        let whiteY = xyzF(from: white).second
        guard whiteY.isFinite, whiteY > 0 else { throw CIE2006ChromaticityError.invalidSpectrum }
        let normalization = 100 / whiteY
        let lms = Vector3(
            first: integrals[0].object * normalization,
            second: integrals[1].object * normalization,
            third: integrals[2].object * normalization
        )
        let xyz = xyzF(from: lms)
        guard [xyz.first, xyz.second, xyz.third].allSatisfy(\.isFinite) else {
            throw CIE2006ChromaticityError.invalidSpectrum
        }
        guard xyz.first != 0 || xyz.second != 0 || xyz.third != 0 else {
            throw CIE2006ChromaticityError.zeroSignal
        }
        guard let point = point(from: xyz) else { throw CIE2006ChromaticityError.invalidSpectrum }
        let whiteLMS = Vector3(first: white.first * normalization, second: white.second * normalization,
                               third: white.third * normalization)
        return MeasurementResult(lms: lms, whiteLMS: whiteLMS, xyzF: xyz, point: point,
                                 wavelengthRange: .init(start: lower, end: upper))
    }

    private static func integrate(
        cone: PhysiologicalCone, reflectance: [SpectralSample], illuminant: [SpectralSample],
        range: ClosedRange<Double>
    ) -> (object: Double, white: Double) {
        let values = cone.sensitivities
        let upper = min(range.upperBound, CIE2006LMSData.firstWavelength + Double(values.count - 1) * CIE2006LMSData.interval)
        let lower = range.lowerBound
        guard lower < upper else { return (0, 0) }
        // CIE metadata specifies linear interpolation and zero extrapolation of each
        // fundamental. Integrating S only through 615 nm avoids inventing its missing tail.
        let nodes = Set(
            [lower, upper] + reflectance.map(\.wavelength) + illuminant.map(\.wavelength)
                + values.indices.map { CIE2006LMSData.firstWavelength + Double($0) * CIE2006LMSData.interval }
        ).filter { $0 >= lower && $0 <= upper }.sorted()
        var object = 0.0
        var white = 0.0
        for (a, b) in zip(nodes, nodes.dropFirst()) {
            let midpoint = (a + b) / 2
            // Between knots all three factors are linear. Simpson's rule integrates
            // their cubic product without altering the original tabulated channel gains.
            for (wavelength, weight) in [(a, 1.0), (midpoint, 4.0), (b, 1.0)] {
                let weightedSource = interpolated(illuminant, at: wavelength)
                    * sensitivity(values, at: wavelength) * (b - a) * weight / 6
                white += weightedSource
                object += weightedSource * interpolated(reflectance, at: wavelength)
            }
        }
        return (object, white)
    }

    private static func sensitivity(_ values: [Double], at wavelength: Double) -> Double {
        let index = (wavelength - CIE2006LMSData.firstWavelength) / CIE2006LMSData.interval
        guard index >= 0, index <= Double(values.count - 1) else { return 0 }
        let lower = Int(index.rounded(.down))
        let upper = min(lower + 1, values.count - 1)
        return values[lower] + (values[upper] - values[lower]) * (index - Double(lower))
    }

    /// Callers have validated order and clipped to the shared measured/source range.
    private static func interpolated(_ samples: [SpectralSample], at wavelength: Double) -> Double {
        var lower = 0
        var upper = samples.count - 1
        while lower + 1 < upper {
            let middle = (lower + upper) / 2
            if samples[middle].wavelength <= wavelength { lower = middle } else { upper = middle }
        }
        let fraction = (wavelength - samples[lower].wavelength) / (samples[upper].wavelength - samples[lower].wavelength)
        return samples[lower].value + (samples[upper].value - samples[lower].value) * fraction
    }
}
